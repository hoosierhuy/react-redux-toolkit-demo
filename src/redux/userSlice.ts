// Libraries
import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { v4 as uuidv4 } from 'uuid'

// App Level
import { RootState } from './store'
import { UserState } from '../types'

type initialStateType = {
  userList: UserState[]
}

const userList: UserState[] = [
  {
    id: '1', // I'm casting this id to a single char because the unit test can't work with uuidv4
    username: 'Huy',
    favoriteColor: 'red',
  },
  {
    id: uuidv4(),
    username: 'Raphael',
    favoriteColor: 'green',
  },
]

const initialState: initialStateType = {
  userList,
}

export const userSlice = createSlice({
  name: 'user',
  initialState,
  reducers: {
    addNewUser: (state, action: PayloadAction<UserState>) => {
      state.userList.push(action.payload)
    },

    deleteUser: (state, action: PayloadAction<{ id: string }>) => {
      state.userList = state.userList.filter(
        (user) => user.id !== action.payload.id
      )
    },
  },
})

export const { addNewUser, deleteUser } = userSlice.actions

export const selectUserList = (state: RootState) => state.user.userList

export default userSlice.reducer
