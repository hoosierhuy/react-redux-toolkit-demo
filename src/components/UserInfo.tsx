// Chakra UI
import { Button, Box, Heading, Text } from '@chakra-ui/react'

// App Level
import { useAppDispatch } from '../hooks'
import { deleteUser } from '../redux/userSlice'

const UserInfo = ({
  username,
  favoriteColor,
  id,
  ...rest
}: {
  username: string | undefined
  favoriteColor: string | undefined
  id: string
}) => {
  const dispatch = useAppDispatch()

  return (
    <Box
      p={3}
      height="4rem"
      width="23.375rem"
      justifyContent="space-between"
      d="flex"
      borderWidth="1px"
      borderColor={favoriteColor}
      {...rest}
    >
      <Box d="flex" flexDirection="column">
        <Text fontSize="l">{username}</Text>
      </Box>

      <Button
        colorScheme="black"
        backgroundColor="gray.200"
        variant="outline"
        borderRadius="0"
        onClick={() => dispatch(deleteUser({ id }))}
      >
        Delete
      </Button>
    </Box>
  )
}

export default UserInfo
