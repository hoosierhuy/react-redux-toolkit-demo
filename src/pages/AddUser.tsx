// React
import { useState } from 'react'
import { Link } from 'react-router-dom'

// Chakra UI
import {
  Box,
  Button,
  Flex,
  FormControl,
  FormLabel,
  Input,
  Select,
} from '@chakra-ui/react'

// Other libraries
import { v4 as uuidv4 } from 'uuid'

// App level
import { useAppDispatch } from '../hooks'
import { addNewUser } from '../redux/userSlice'

const AddUser = () => {
  const dispatch = useAppDispatch()

  const [username, setUsername] = useState<string | undefined>('')

  const colorsArr = [
    'Please select favorite color',
    'red',
    'orange',
    'blue',
    'purple',
  ]
  const [favoriteColor, setFavoriteColor] = useState<string | undefined>('')

  const handleOnSubmit = () => {
    if (username)
      dispatch(addNewUser({ favoriteColor, username, id: uuidv4() }))

    setFavoriteColor('')
    setUsername('')
  }

  return (
    <Flex
      height="50vh"
      justifyContent="center"
      alignItems="center"
      flexDirection="column"
      color="black"
    >
      <Box>
        <FormControl isRequired>
          <FormLabel color="black">User Name</FormLabel>
          <Input
            width="23.375rem"
            height="2.5rem"
            borderRadius="0"
            borderColor="#888888"
            value={username}
            isRequired
            placeholder="Enter user name"
            onChange={(evt) => setUsername(evt.currentTarget.value)}
          />
          <FormLabel color="black" marginTop={4}>
            Favorite Color
          </FormLabel>
          <Select
            width="23.375rem"
            height="2.5rem"
            borderRadius="0"
            borderColor="#888888"
            isRequired
            value={favoriteColor}
            onChange={(evt) => setFavoriteColor(evt.target.value)}
          >
            {colorsArr.map((color) => (
              <option key={color} value={color}>
                {color}
              </option>
            ))}
          </Select>
        </FormControl>

        <Button
          width="23.375rem"
          height="2.5rem"
          borderRadius="0"
          marginTop={8}
          backgroundColor="#FFC700"
          type="submit"
          onClick={handleOnSubmit}
        >
          Submit
        </Button>

        <Box d="flex" flexDirection="row" justifyContent="center" marginTop={8}>
          <Link to="/userlist">View User List</Link>
        </Box>
      </Box>
    </Flex>
  )
}

export default AddUser
