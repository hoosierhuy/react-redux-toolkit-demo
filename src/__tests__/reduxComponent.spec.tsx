import { render, screen, fireEvent, RenderResult } from '@testing-library/react'

import { Provider } from 'react-redux'

import { MemoryRouter } from 'react-router-dom'

import UserInfo from '../components/UserInfo'
import AddUser from '../pages/AddUser'

import { store } from '../redux/store'
import { UserState } from '../types'

const renderUser = (user: UserState): RenderResult =>
  render(
    <Provider store={store}>
      <UserInfo
        username={user.username}
        favoriteColor={user.favoriteColor}
        id={user.id}
      />
    </Provider>
  )

const renderAddUser = (): RenderResult =>
  render(
    <Provider store={store}>
      <MemoryRouter>
        <AddUser />
      </MemoryRouter>
    </Provider>
  )

const getAUser = (userId: string): UserState => {
  const user = store.getState().user.userList.find((user) => user.id === userId)
  expect(user).not.toBeUndefined()
  return user as UserState
}

test('Renders UserInfo', () => {
  const user = getAUser('1')
  const { asFragment } = renderUser(user)
  expect(asFragment()).toMatchSnapshot()
})

test('AddUser page', () => {
  renderAddUser()
  const initialLength = store.getState().user.userList.length

  let usernameInput = screen.getByPlaceholderText('Enter user name')

  expect(usernameInput).toBeInTheDocument()

  fireEvent.change(usernameInput, {
    target: { value: 'Please select favorite color' },
  })

  let submitButton = screen.getByText('Submit')

  fireEvent.click(
    submitButton,
    new MouseEvent('click', {
      bubbles: true,
      cancelable: true,
    })
  )

  let user = store.getState().user.userList.length
  expect(user).toBeGreaterThan(initialLength)
})
