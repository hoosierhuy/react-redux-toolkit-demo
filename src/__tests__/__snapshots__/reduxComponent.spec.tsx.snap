// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders UserInfo 1`] = `
<DocumentFragment>
  <div
    class="css-1fdecp7"
  >
    <div
      class="css-j7qwjs"
    >
      <p
        class="chakra-text css-16mom77"
      >
        Huy
      </p>
    </div>
    <button
      class="chakra-button css-1iz9f5j"
      type="button"
    >
      Delete
    </button>
  </div>
</DocumentFragment>
`;
